// @foal/cli/lib/generate/utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.installDependencies = exports.logger = exports.getNames = exports.mkdirIfDoesNotExist = exports.rmDirAndFilesIfExist = exports.initGitRepo = void 0;
var init_git_repo_1 = require("./init-git-repo");
Object.defineProperty(exports, "initGitRepo", { enumerable: true, get: function () { return init_git_repo_1.initGitRepo; } });
var rm_dir_and_files_if_exist_1 = require("./rm-dir-and-files-if-exist");
Object.defineProperty(exports, "rmDirAndFilesIfExist", { enumerable: true, get: function () { return rm_dir_and_files_if_exist_1.rmDirAndFilesIfExist; } });
var mkdir_if_does_not_exist_1 = require("./mkdir-if-does-not-exist");
Object.defineProperty(exports, "mkdirIfDoesNotExist", { enumerable: true, get: function () { return mkdir_if_does_not_exist_1.mkdirIfDoesNotExist; } });
var get_names_1 = require("./get-names");
Object.defineProperty(exports, "getNames", { enumerable: true, get: function () { return get_names_1.getNames; } });
var logger_1 = require("./logger");
Object.defineProperty(exports, "logger", { enumerable: true, get: function () { return logger_1.logger; } });
var install_dependencies_1 = require("./install-dependencies");
Object.defineProperty(exports, "installDependencies", { enumerable: true, get: function () { return install_dependencies_1.installDependencies; } });
