// @foal/typeorm/lib/index.js
"use strict";
/**
 * FoalTS
 * Copyright(c) 2017-2022 Loïc Poullain <loic.poullain@centraliens.net>
 * Released under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConcreteSessionStore = exports.TypeORMStore = exports.DatabaseSession = exports.UserWithPermissions = exports.Permission = exports.Group = void 0;
var entities_1 = require("./entities");
Object.defineProperty(exports, "Group", { enumerable: true, get: function () { return entities_1.Group; } });
Object.defineProperty(exports, "Permission", { enumerable: true, get: function () { return entities_1.Permission; } });
Object.defineProperty(exports, "UserWithPermissions", { enumerable: true, get: function () { return entities_1.UserWithPermissions; } });
var typeorm_store_service_1 = require("./typeorm-store.service");
Object.defineProperty(exports, "DatabaseSession", { enumerable: true, get: function () { return typeorm_store_service_1.DatabaseSession; } });
Object.defineProperty(exports, "TypeORMStore", { enumerable: true, get: function () { return typeorm_store_service_1.TypeORMStore; } });
Object.defineProperty(exports, "ConcreteSessionStore", { enumerable: true, get: function () { return typeorm_store_service_1.TypeORMStore; } });
